2.2.0~alpha blog post (#92)
ocaml/platform-blog: opam-2-2-0-alpha.md

-----
title: "opam 2.2.0 alpha is ready!"
authors: [
  "Raja Boujbel - OCamlPro" {"mailto:raja.boujbel(à)ocamlpro.com"}
  "Louis Gesbert - OCamlPro" {"mailto:louis.gesbert(à)ocamlpro.com"}
]
date: "2023-06-30"
--BODY--

We are happy to announce the alpha release of opam 2.2.0. It contains numerous
fixes, enhancements, updates. You can view the full list in the [release
note](https://github.com/ocaml/opam/releases/tag/2.2.0-alpha).

This version is an alpha, so we invite users to test it to spot previously
unnoticed bugs and work towards a stable release.

## Windows Support

Opam 2.2 comes with native Windows compatibility. You can now use opam from
your preferred Windows terminal! We rely on the [Cygwin](https://cygwin.com/)
UNIX-like environment for Windows as a compatibility layer, but it is possible
for a package to generate native executables.

The main opam repository is not Windows compatible at the moment, but existing
work on a [compatible
repository](https://github.com/ocaml-opam/opam-repository-mingw) (originally

> from [@fdopen](https://github.com/fdopen)) and [32/64 bit mingw-w64 packages](https://github.com/dra27/opam-repository) (by [@dra27](https://github.com/dra27)) is in the process of being merged.

Author: ```suggestion
from [@fdopen](https://github.com/fdopen)) and [32/64 bit mingw-w64 packages](https://github.com/dra27/opam-repository) (by [@dra27](https://github.com/dra27)) is in the process of being merged. By the release, main opam repository we expect opam repository to be Windows compliant.
```
@dra27 ?

### How to Test opam on Windows

This alpha requires a preexisting Cygwin installation. Support for full
management of a local Cygwin environment inside of opam (so that it's as
transparent as possible) is queued already and should be available

> in 2.2.0~alpha2 as another option.

Author: It is not meant to be the default, if there is an already installation, we propose to user.

> 1. Check that you have all dependencies installed: Autoconf, Make, MinGW, or MSVC GCC

Author: ```suggestion
1. Check that you have all dependencies installed:
  * `autoconf`, `make`
  * MinGW compilers: `mingw64-x86_64-gcc-g++`, `mingw64-i686-gcc-g++`
  * Or If you want to use the MSVC port of OCaml, you'll need to install Visual Studio or Visual Studio Build Tools
```

2. Download & extract the [opam archive](https://github.com/ocaml/opam/releases/download/2.2.0-alpha/opam-full-2.2.0-alpha.tar.gz)

> 3. In the directory launch `make cold [OCAML_PORT=mingw64|mingw|msvc64|msvc|auto]`

Author: I was worried that it would choose default cygwin compiler
```suggestion
3. In the directory launch `make cold`
```

> 3. In the directory launch `make cold [OCAML_PORT=mingw64|mingw|msvc64|msvc|auto]`

Reviewer: Hopefully people won't have installed it - that's the `gcc-core` package in Cygwin (which isn't picked up by default).

4. A coffee later, you now have an opam executable!

> 5. Start your preferred Windows terminal (cmd or PowerShell), and initialise opam with the Windows _sunset_ repository:

Author: cf comment :arrow_up:

  * `opam init https://github.com/ocaml-opam/opam-repository-mingw`

> From here, you can try to install sunset repository packages. If any bug is

Author: ```suggestion
From here, you can try to install [sunset repository](https://discuss.ocaml.org/t/sunsetting-opam-repository-mingw/11632) packages. If any bug is
```

found, please [submit an
issue](https://github.com/ocaml-opam/opam-repository-mingw#what-do-i-do-when-things-are-broken).
It will help opam repository maintainers to add Windows repository packages
into the main repository.

> Hint: if you use the MinGW compiler, don't forget to add to your `PATH` the path

> to `libc` dlls.

Author: It is needed to compile/launch opam itself.

> to `libc` dlls.

Reviewer: Ah, good point - we could instead have `make cold` pass `--with-private-runtime` to configure. The only caveat is that then you have to copy `opam.exe` and the generated runtime directory (this is what fdopen's opam64 distribution uses).

> > Hint: if you use the MinGW compiler, don't forget to add to your `PATH` the path
> to `libc` dlls.

Author: ```suggestion
> Hint: if you use the MinGW compiler, don't forget to add to your `PATH` the path to `libc` dlls (usually `C:\cygwin64\usr\x86_64-w64-mingw32\sys-root\mingw\bin`). Or compile opam with `make cold CONFIGURE_ARGS=--with-private-runtime`, and if you change opam location, don't forget to copy `Opam.Runtime.amd64` (or `Opam.Runtime.i386`) with it.
```

## Recursive Pin

Recursive pinning allows opam to lookup opam files in
subdirectories in different ways:

* `--recursive` option, opam goes through subdirectories to look for opam
  files and pin them to their respective subdirectories 
* `--subpath <path>` option, opam goes to the given subdirectory to find the opam
  file and pin the corresponding package to this subdirectory 

> * These options can be combined

Reviewer 2: When installing or pinning a package using `opam install` or `opam pin`, opam normally only looks for opam files at the root of the installed package. With recursive pinning, you can now instruct opam to also look for `.opam` files in subdirectories, while maintaining the correct relationship between the `.opam` files and the package root for versioning and build purposes.

Recursive pinning is used with the following options to `opam pin` and `opam install`:
 * With `--recursive`, opam will look for `.opam` files recursively in all subdirectories.
 * With `--subpath <path>`, opam will only look for `.opam` files in the subdirectory `<path>`.

The two options can be combined: for instance, if your opam packages are stored as a deep hierarchy in the `opam` subdirectory of your project, give `opam pin . --recursive --subpath opam` a try!

Reviewer 3: Thanks this seems much clearer

Reviewer: A thought - is it possibly clearer to demonstrate `--subpath` _first_ and then `--recursive`? IIUC we could show say a manual sub-path pin of `bar/bar.opam` in the example and then do the recursive pinning afterwards (and it will add another)?

> You can use these options with `opam pin` and `opam install`.

Author: ```suggestion
You can use these options with `opam pin`, `opam install`, and ``opam remove`.
```

> You can use these options with `opam pin`, `opam install`, and ``opam remove`.

Reviewer: ```suggestion
You can use these options with `opam pin`, `opam install`, and `opam remove`.
```

```
$ tree .
.
├── ba
│   └── z
│       └── z.opam
├── bar
│   └── bar.opam
└── foo.opam

> $ opam pin --recursive . -n

Reviewer: I think it's always clearer in blogs/docs to use the full option names:
```suggestion
$ opam pin --recursive . --no-action
```

└── user-setup.0.7
```

> Special thanks to [@cannorin](htpps://github.com/@cannorin) for collaborating on this feature.

## Recommended Development Tools

There is now a way for a project maintainer to share their project development
tools: the `with-dev-setup` dependency flag. It is used in the same way as
`with-doc` and `with-test`: by adding a `{with-dev-setup}` filter after a
dependency. It will be ignored when installing normally, but it's pulled in when the
package is explicitely installed with the `--with-dev-setup` flag specified on
the command line. The variable is also resolved in the `post-messages:` field
to allow maintainers to share more informations about that setup.

This is typically useful for tools that are required for bootstrapping or
regenerating artifacts.

For example
```
opam-version: "2.0"
depends: [
  "ocaml"
  "dune"
  "ocp-indent" {with-dev-setup}
]
build: [make]
install: [make "install"]
post-messages:
[ "Thanks for installing the package"
  "and its tool dependencies too, it will help for your futur PRs" {with-dev-setup} ]
```

## Software Heritage Binding

[Software Heritage](https://www.softwareheritage.org) is a project that aims to

> archive all software source code in existence. This is done by first collecting

Reviewer: There's "first" here, but not then the "second" 🙂 I think the second part (which we're not doing yet) is having the tooling to add the SWHIDs in opam-repository. Is it worth mentioning that the aim is for that to follow the release of 2.2 (and then go on to say that you can manually add them to your packages right now, as shown below)?

Author: ```suggestion
archive all software source code in existence. This is done by collecting
```
There is no second indeed, it was the introduction of the whole process. The `first` referred to the `from there` below.

SWHID. If an SWHID URL is present in an opam file, the fallback can be
activated.

To keep backwards compatibility of opam files, we added a specific
Software Heritage URL syntax to the `url.mirrors:` field, which is used to specify mirrors of the main URL. Opam 2.2.+ understands this specific syntax as a Software
Heritage fallback URL: `https://swhid.opam.ocaml.org/<SWHID>`.

```
url {
  src: "https://faili.ng/url.tar.gz"
  checksum: "sha512=e2146c1d7f53679fd22df66c9061b5ae4f8505b749513eedc67f3c304f297d92e54f5028f40fb5412d32c7d7db92592eacb183128d2b6b81d10ea716b7496eba"
  mirrors: [
    "https//failli.ng/mirror.tar.gz"
    "https://swhid.opam.ocaml.org/swh:1:dir:9f2be900491e1dabfc027848204ae01aa88fc71d"
  ]
}
```

To add a Software Heritage fallback URL to your package, use the
[`swhid`](https://github.com/OCamlPro/swhid) library. Specifically the
[`Compute.directory_identifier_deep`](https://github.com/OCamlPro/swhid/blob/master/src/compute.mli#L12)
function:

1. Download opam package archive
2. Extract the archive
3. Compute SWHID with `Compute.directory_identifier_deep`. You can use this oneliner in the directory:
```ocaml

> ocaml -e '#use "topfind";; #require "digestif.ocaml";; #require "swhid";; Swhid_core.Object.pp Format.std_formatter (Result.get_ok (Swhid.Compute.directory_identifier_deep "."))'

Reviewer: Is there already an issue for having a command to do this in opam itself? (sorry, we may have discussed this ages ago!)

Reviewer: Yeah, agreed - or perhaps the swhid package could have a binary to do the same?

Author: It is planned ;)

```

> Special thanks to [@zapashcanon](https://github.com/@zapashcanon) for collaborating on this feature

## Formula (Experimental)

It is now possible to leverage the full expressivity of package dependency
formulas from the command line during switch creation and package operations.

> 

Author: ````suggestion

It is possible to create a switch using a formula, for example:
 with `ocaml-variant` or `ocaml-system`, excluding `ocaml-base-compiler`:

```
opam switch create ocaml --formula '"ocaml-variants" {>= "4.14.1"} | "ocaml-system"'
```

This syntax is brought to install commands.
````

For example, while installing a package, let's say `genet`, you can specify
that you want to install either `conf-mariadb & mariadb` or `conf-postgresql`:

```
opam install genet --formula '["mysql" ("conf-mariadb" & "mariadb" | "conf-postgresql")]'
```

Or create a switch with `ocaml-variant` or `ocaml-system`, excluding `ocaml-base-compiler`:

```

> opam switch create ocaml --formula '"ocaml-variants" {>= "4.14.1"} | "ocaml-system"'

Reviewer: This part isn't new - is it worth turning this around, so starting with the `opam switch create --formula` and then explaining how that's been extended to `opam install` (IIRC you can't mix `--packages` and `--formula` in `opam switch create`? That might be worth noting that the version in `opam install` is a slightly more powerful UI?)

Author: cf comment :arrow_up:

> Or create a switch with `ocaml-variant` or `ocaml-system`, excluding `ocaml-base-compiler`:
> 
> ```
> opam switch create ocaml --formula '"ocaml-variants" {>= "4.14.1"} | "ocaml-system"'
> ```

Author: ```suggestion
```

## New Options
Here are several of new options (possibly scripts breaking changes are marked with ✘):
  * `opam pin --current` to fix a package to its current state (disabling pending reinstallations or removals from the repository)
  * `opam pin remove --all` to remove all the pinned packages from a switch
  * Allow `opam pin remove` to take a package (`<pkg>.<version>`) as argument
  * `opam exec --no-switch` to remove opam environment from launched command
  * `opam source --no-switch`
  * `opam clean --untracked` to remove untracked files interactively
  * `opam switch -` that goes back to the previously selected global switch
  * Add `opam admin add-constraint <cst> --packages` to select a subset of packages to apply constraints

>   * ✘ Change `--base` into `--invariant`. `opam switch` _compiler_ column now contains installed packages that verifier invariant formula, and empty synopsis shows switch invariant.

Reviewer: It's very easy for me to say, but I think these would be better with a one-liner example added or description of what didn't work before! 🙂 Something like:
- `opam pin --current` - sample tree and explanation of the resulting state of the switch
- `opam pin remove foo.version` now removes the pin on `pkg` (i.e. describe the change rather than the option)
- `opam exec --no-switch` can probably be demonstrated quite nicely with an `env | grep -F opam` command or something to show that it's empty?
- `opam source --no-switch` - motivation is that you don't therefore need to have created any switch
- `opam switch -` - possibly worth referencing the same feature in git?
- etc

The change for base packages I would say either wants a section or to be omitted (remembering the pain that it was, I definitely see the desire to have it mentioned!)

-----
title: "opam 2.2.0 alpha is ready!"

> authors: [
>   "Raja Boujbel - OCamlPro" {"mailto:raja.boujbel(à)ocamlpro.com"}
>   "Louis Gesbert - OCamlPro" {"mailto:louis.gesbert(à)ocamlpro.com"}
> ]

Author: Usually (previous posts) we set in the authors the person(s) who wrote the post, not necessarily the reviewers. But it was maybe a mistake.
```suggestion
authors: [
  "David Allsopp - Tarides" {"mailto:david(at)tarides"}
  "Léo Andrès - OCamlPro" {"mailto:leo.andres(at)ocamlpro.com"}
  "Raja Boujbel - OCamlPro" {"mailto:raja.boujbel(at)ocamlpro.com"}
  "Basile Clément - OCamlPro" {"mailto:basile.clement(at)ocamlpro.com"}
  "Louis Gesbert - OCamlPro" {"mailto:louis.gesbert(at)ocamlpro.com"}
  "Dario Pinto" - OCamlPro" {"mailto:dario.pinto(at)ocamlpro.com"}
  "Christine Rose - Tarides" {"mailto:christine(at)tarides.com"}
  "Riku Silvola - Tarides" {"mailto:riku(at)tarides.com"}
]
```

> ]

Reviewer 3: The point of `(à)` was to lightly obfuscate the addresses against scrapper bots; I think it serves little purpose to use `(at)` and not just `@`. Then again, it might have been useless already...

> ]

Reviewer: Yeah, surely somebody's AI by now is accurately interpreting email addresses. FWIW, in nearly all (all?) those cases we could link GitHub profiles instead, or do they have to be mailto?

> date: "2023-06-28"

Author: ```suggestion
date: "2023-06-30"
```

--BODY--

> We are happy to announce the alpha release of opam 2.2.0. It contains numerous
> fixes, enhancements, updates. You can view the full list in the [release
> note](https://github.com/ocaml/opam/releases/tag/2.2.0-alpha).

Author: ```suggestion
We are happy to announce the alpha release of opam 2.2.0. It contains numerous
fixes, enhancements, and updates; including much-improved Windows support, addressing one of the most important pain points [identified by the OCaml community](https://discuss.ocaml.org/t/what-are-the-biggest-reasons-newcomers-give-up-on-ocaml/10958). You can view the full list of changes in the [release
note](https://github.com/ocaml/opam/releases/tag/2.2.0-alpha).
This alpha release is a significant milestone for the opam dev team which remains a long-term collaboration between OCamlPro ([@rjbou](https://github.com/rjbou), [@AltGr](https://github.com/AltGr)) and Tarides ([@dra27](https://github.com/dra27), [@kit-ty-kate](https://github.com/kit-ty-kate)), and the work of tens of community contributors.
```

> We are happy to announce the alpha release of opam 2.2.0. It contains numerous
> fixes, enhancements, updates. You can view the full list in the [release
> note](https://github.com/ocaml/opam/releases/tag/2.2.0-alpha).

Reviewer 3: ```suggestion
We are happy to announce the alpha release of opam 2.2.0. It contains numerous
fixes, enhancements, and updates; including much-improved Windows support, addressing one of the most important pain points [identified by the OCaml community](https://discuss.ocaml.org/t/what-are-the-biggest-reasons-newcomers-give-up-on-ocaml/10958). You can view the full list of changes in the [release
note](https://github.com/ocaml/opam/releases/tag/2.2.0-alpha).

This alpha release is a significant milestone, brought together by Raja Boujbel after years of work from the opam dev team (Raja Boujbel, David Allsopp, Kate Deplaix, Louis Gesbert, in a united OCamlPro/Tarides collaboration) with the help of many community contributors. We also thank JaneStreet for their continued sponsorship.
```

This version is an alpha, we invite users to test it to spot previously
unnoticed bugs and get towards a stable release.

## Windows Support

Opam 2.2 comes with native Windows compatibility. You can now use opam from
your preferred Windows terminal!  We rely on the [Cygwin](https://cygwin.com/)
UNIX-like environment for Windows as a compatibility layer, but it is possible
for package to generate native executables.

The main opam repository is not Windows compliant at the moment, but existing
work on a [compatible
repository](https://github.com/ocaml-opam/opam-repository-mingw) (originally
from @fdopen) is in the process of being merged.

> ### How to test opam on Windows

Reviewer 4: So sorry for the misunderstanding. That's exactly what I mean. It's always lowercase like "opam" unless it's at the beginning of a sentence. Even in titles/headings/subheadings it's "opam."

switch.

```
$ opam tree cppo
cppo.1.6.9
├── base-unix.base
├── dune.3.8.2 (>= 1.10)
│   ├── base-threads.base
│   ├── base-unix.base [*]
│   └── ocaml.4.14.1 (>= 4.08)
│       ├── ocaml-base-compiler.4.14.1 (>= 4.14.1~ & < 4.14.2~)
│       └── ocaml-config.2 (>= 2)
│           └── ocaml-base-compiler.4.14.1 (>= 4.12.0~) [*]
└── ocaml.4.14.1 (>= 4.02.3) [*]
```

It can also display a reverse-dependency tree (through `opam why`, which is an
alias to `opam tree --rev-deps`), that is useful to examine how dependency
versions get constrained.

```
$ opam why cmdliner
cmdliner.1.2.0
├── (>= 1.1.0) b0.0.0.5
│   └── (= 0.0.5) odig.0.0.9
├── (>= 1.1.0) ocp-browser.1.3.4
├── (>= 1.0.0) ocp-indent.1.8.1
│   └── (>= 1.4.2) ocp-index.1.3.4
│       └── (= version) ocp-browser.1.3.4 [*]
├── (>= 1.1.0) ocp-index.1.3.4 [*]
├── (>= 1.1.0) odig.0.0.9 [*]
├── (>= 1.0.0) odoc.2.2.0
│   └── (>= 2.0.0) odig.0.0.9 [*]
├── (>= 1.1.0) opam-client.2.2.0~alpha
│   ├── (= version) opam.2.2.0~alpha
│   └── (= version) opam-devel.2.2.0~alpha
├── (>= 1.1.0) opam-devel.2.2.0~alpha [*]
├── (>= 0.9.8) opam-installer.2.2.0~alpha
└── user-setup.0.7
```

> > Special thanks to @cannorin for collaborating on this feature

Author: ```suggestion
> Special thanks to [@cannorin](htpps://github.com/@cannorin) for collaborating on this feature
```

`url.mirrors:` (that is used to specify mirrors of the main url) a specific
Software Heritage URL syntax that is understood by opam 2.2.+ as a Software
Heritage fallback url: `https://swhid.opam.ocaml.org/<SWHID>`.

```
url {
  src: "https://faili.ng/url.tar.gz"
  checksum: "sha512=e2146c1d7f53679fd22df66c9061b5ae4f8505b749513eedc67f3c304f297d92e54f5028f40fb5412d32c7d7db92592eacb183128d2b6b81d10ea716b7496eba"
  mirrors: [
    "https//failli.ng/mirror.tar.gz"
    "https://swhid.opam.ocaml.org/swh:1:dir:9f2be900491e1dabfc027848204ae01aa88fc71d"
  ]
}
```

To add a Software Heritage fallback URL to your package, use the
[`swhid`](https://github.com/OCamlPro/swhid) library. Specifically the
[`Compute.directory_identifier_deep`](https://github.com/OCamlPro/swhid/blob/master/src/compute.mli#L12)
function:

1. Download opam package archive
2. Extract the archive
3. Compute SWHID with `Compute.directory_identifier_deep`.  You can use this oneliner in the directory:
```ocaml
ocaml -e '#use "topfind";; #require "digestif.ocaml";; #require "swhid";; Swhid_core.Object.pp Format.std_formatter (Result.get_ok (Swhid.Compute.directory_identifier_deep "."))'
```

> > Special thank to @zapashcanon for collaborating on this feature

Author: ```suggestion
> Special thanks to [@zapashcanon](https://github.com/@zapashcanon) for collaborating on this feature
```

## Formula (experimental)

It is now possible to leverage the full expressivity of package dependency
formulas from the command-line during switch creation and package operations.

For example, while installing a package, let's say `genet`, you can specify
that you want to install either `conf-mariadb & mariadb` or `conf-postgresql`

```
opam install genet --formula '["mysql" ("conf-mariadb" & "mariadb" | "conf-postgresql")]'
```

Or create a switch with `ocaml-variant` or `ocaml-system`, excluding `ocaml-base-compiler`

```
opam switch create ocaml --formula '"ocaml-variants" {>= "4.14.1"} | "ocaml-system"'
```

## New options

> And a bunch of new options:

Author: ```suggestion
Here are several of new options (possibly scripts breaking changes are marked with ✘):
```

  * `opam pin --current` to fix a package to its current state (disabling pending reinstallations or removals from the repository)
  * `opam pin remove --all` to remove all the pinned packages from a switch
  * Allow `opam pin remove` to take a package (`<pkg>.<version>`) as argument
  * `opam exec --no-switch` to remove opam environment from launched command
  * `opam source --no-switch`
  * `opam clean --untracked` to remove untracked files interactively
  * `opam switch -` that goes back to the previously selected global switch
  * Add `opam admin add-constraint <cst> --packages` to select a subset of packages to apply constraints

>   * ✘ Change `--base` into `--invariant`, column name and the content is invariant formula installed dependencies

Reviewer 4: ```suggestion
  * ✘ Change `--base` into `--invariant`. The column name and the content are invariant formula-installed dependencies
```

Author: Oups forgot to add a note on that.

## Try it!

In case you plan a possible rollback, you may want to first backup your
`~/.opam` directory.

The upgrade instructions are unchanged:

> 1. Either from binaries: run
> 
>     ```
>     bash -c "sh <(curl -fsSL https://raw.githubusercontent.com/ocaml/opam/master/shell/install.sh) --version 2.2.0~alpha"
>     ```
> 
>     or download manually from [the Github "Releases" page](https://github.com/ocaml/opam/releases/tag/2.2.0-alpha) to your PATH.
> 
> 2. Or from source, manually: see the instructions in the [README](https://github.com/ocaml/opam/tree/2.2.0-alpha#compiling-this-repo).
> 
> 
> You should then run:

Reviewer 4: ```suggestion
1. From binaries: run
    ```
    bash -c "sh <(curl -fsSL https://raw.githubusercontent.com/ocaml/opam/master/shell/install.sh) --version 2.2.0~alpha"
    ```
Or download manually from [the Github "Releases" page](https://github.com/ocaml/opam/releases/tag/2.2.0-alpha) to your PATH.

2. From source, manually: see the instructions in the [README](https://github.com/ocaml/opam/tree/2.2.0-alpha#compiling-this-repo).

Then run:
```

There appears to be something going on with the code block here, but I can't find the extra ``` (unless I'm just not seeing it correctly. I don't have a preview, but the "or download..." and "2. From source..." are in blue, then the 

```
opam init --reinit -ni
```
is in white on my screen. (If that makes sense)

```
opam init --reinit -ni
```

Please report any issues to [the bug-tracker](https://github.com/ocaml/opam/issues).

> Thanks for trying this new release out, and hoping you will enjoy the new features!

Reviewer 4: Minor suggestions: "Finally, we want to recognise everyone for their contribution to getting this alpha release over the line, especially OCamlPro and Tarides for their sponsorship of the project. Thank you!"

British spelling of recognise and replaced "and" with a comma

> Thanks for trying this new release out, and hoping you will enjoy the new features!

Reviewer 4: Also: "Thanks for trying this new release out, and we're hoping you will enjoy the new features!

(add "we're" between "and" and "hoping")
